File: class5.2/MegaDataPool.cpp
#include "MegaDataPool.h"

MegaDataPool::MegaDataPool(int size) : poolSize(size), usedCount(0) {
    pool.resize(size);
    available.resize(size, true);
}

MegaDataPool& MegaDataPool::getInstance(int size) {
    static MegaDataPool instance(size);
    return instance;
}

MegaData* MegaDataPool::acquire() {
    std::lock_guard<std::mutex> lock(mtx);

    for (int i = 0; i < poolSize; ++i) {
        if (available[i]) {
            available[i] = false;
            usedCount++;
            return &pool[i];
        }
    }
    return nullptr;
}

void MegaDataPool::release(MegaData* megadata) {
    std::lock_guard<std::mutex> lock(mtx);

    for (int i = 0; i < poolSize; ++i) {
        if (&pool[i] == megadata) {
            available[i] = true;
            usedCount--;
            pool[i].reset();
            break;
        }
    }
}

int MegaDataPool::size() const {
    return poolSize;
}

int MegaDataPool::usedSize() const {
    return usedCount;
}

File: home5.1/MegaDataPool.java
import java.util.ArrayList;
import java.util.List;

public class MegaDataPool {
    private List<MegaData> pool;
    private List<Boolean> available;
    private int poolSize;
    private int usedCount;

    public MegaDataPool(int size) {
        poolSize = size;
        usedCount = 0;
        pool = new ArrayList<>(size);
        available = new ArrayList<>(size);
        for (int i = 0; i < size; i++) {
            pool.add(new MegaData());
            available.add(true);
        }
    }

    public synchronized MegaData acquire() {
        for (int i = 0; i < poolSize; i++) {
            if (available.get(i)) {
                available.set(i, false);
                usedCount++;
                return pool.get(i);
            }
        }
        return null;
    }

    public synchronized void release(MegaData megadata) {
        int index = pool.indexOf(megadata);
        if (index != -1 && !available.get(index)) {
            available.set(index, true);
            usedCount--;
            megadata.reset();
        }
    }

    public int size() {
        return poolSize;
    }

    public int usedSize() {
        return usedCount;
    }
}
